vsdownload.py

```diff
@@ -111,117 +126,176 @@ def setPackageSelectionMSVC16(args, packages, userversion, sdk, toolversion, def
+        args.ignore.extend(defaultIgnores)
 
-def setPackageSelectionMSVC15(args, packages, userversion, sdk, toolversion, defaultPackages):
+def setPackageSelectionMSVC15(args, packages, userversion, sdk, toolversion, defaultPackages, defaultIgnores):
     if findPackage(packages, "Microsoft.VisualStudio.Component.VC.Tools." + toolversion, warn=False):
-        args.package.extend(["Win10SDK_" + sdk, "Microsoft.VisualStudio.Component.VC.Tools." + toolversion])
+        appendPackageSelection(args, args.with_msvc, "Microsoft.VisualStudio.Component.VC.Tools." + toolversion)
+        if args.sdk_version == None:
+            args.sdk_version = sdk
     else:
         # Options for toolchains for specific versions. The latest version in
         # each manifest isn't available as a pinned version though, so if that
         # version is requested, try the default version.
         print("Didn't find exact version packages for " + userversion + ", assuming this is provided by the default/latest version")
         args.package.extend(defaultPackages)
+        args.ignore.extend(defaultIgnores)
 
 def setPackageSelection(args, packages):
     if not args.architecture:
         args.architecture = ["host", "x86", "x64", "arm", "arm64"]
     if args.host_arch is not None and "host" in args.architecture:
         args.architecture.append(args.host_arch)
 
+    # Select default packages.
+    if args.msvc_version is not None:
+        if args.with_msvc is None:
+            args.with_msvc = True
+        if args.with_asan is None:
+            args.with_asan = True
+        if args.with_atl is None:
+            args.with_atl = True
+        if args.with_sdk is None:
+            args.with_sdk = True
+    if args.sdk_version is not None:
+        if args.with_sdk is None:
+            args.with_sdk = True
+
+    if args.with_default is None:
+        if args.msvc_version is None and len(args.package) == 0:
+            args.with_default = True
+
+    if args.with_default is not None:
+        for component in ["workload", "msvc", "asan", "sdk", "atl", "dia", "msbuild", "devcmd"]:
+            if getattr(args, "with_" + component) is None:
+                setattr(args, "with_" + component, args.with_default)
+
     # If no packages are selected, install these versionless packages, which
     # gives the latest/recommended version for the current manifest.
-    defaultPackages = ["Microsoft.VisualStudio.Workload.VCTools"]
+    defaultPackages, args.package = args.package, []
+    defaultIgnores, args.ignore = args.ignore, []
+
+    appendPackageSelection(args, args.with_workload, "Microsoft.VisualStudio.Workload.VCTools")
+
     if "x86" in args.architecture or "x64" in args.architecture:
-        defaultPackages.append("Microsoft.VisualStudio.Component.VC.ATL")
+        appendPackageSelection(args, args.with_msvc, "Microsoft.VisualStudio.Component.VC.Tools.x86.x64")
+        appendPackageSelection(args, args.with_asan, "Microsoft.VisualCpp.ASAN.X86")
```

Comment:
If I understand the patch correctly, this ASAN package, and the x86.64 tools above, are packages which previously got pulled in automatically through `Workload.VCTools` above. Now in a total default run, `vsdownload.py --dest /opt/msvc --accept-license`, we'd get them explicitly pulled in both implicitly via the workload and explicitly - is that correct?

In that case, do I understand correctly, that we also could make `--with-workload` default to false, to reduce the default install size, as we now do explicitly select the individual packages that we do need? (I guess we could do that as a separate commit then.)

Reply:
Yes, you are correct. Making `--with-workload` default to false is the goal, but it need more tests to see if any more packages are needed.